### PR TITLE
Scientific log for small numbers

### PR DESCRIPTION
**Pull request by victoria34**
Add minExponent and maxExponent. Check if data of getDecimalExponents is empty.

### CODE REVIEW

File: app/_components/cbio-util.js [hunk 1]
```diff
@@ -666,6 +672,47 @@ cbio.util = (function() {
     return def.promise();
   }
 
+
+  function getDecimalExponents(data){
+
+    // Copy the values, rather than operating on references to existing values
+    if (data === null || data.length < 1) {//if data is null or empty, return data
```
Comment (zhx828) [inline app/_components/cbio-util.js]: You should use `_.isArray(data)` instead of `data === null`

File: app/_components/cbio-util.js [hunk 2]
```diff
@@ -666,6 +672,47 @@ cbio.util = (function() {
...
+    
+    var values = [];
+    _.each(data, function(item) {
+      if (!isNaN(item)) {
+        values.push(Number(item));
+      }
+    });
+
+    // Then sort
+    values.sort(function(a, b) {
```
Comment (zhx828) [inline app/_components/cbio-util.js:694]: The default sorting is a-b. You should use `values.sort()` directly

File: app/_components/cbio-util.js [hunk 3]
```diff
@@ -666,6 +672,47 @@ cbio.util = (function() {
...
+      return a - b;
+    });
+    
+    var minZeros = 0, maxZeros = 0;
+      
+    while (values[0] < 1) {
```
Comment (zhx828) [inline app/_components/cbio-util.js]: Like I said before, `values` is not guaranteed to have any elements

File: app/scripts/views/components/barChart/barChart.js [hunk 4]
```diff
@@ -89,14 +105,19 @@
       if (logScale) {
         chartInst_.x(d3.scale.log().nice()
           .domain([0.7, opts_.maxDomain]));
+      } else if (data_.smallDataFlag) {
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChart.js]: Can't this part be merged with line 115-120?

File: app/scripts/views/components/barChart/barChart.js [hunk 5]
```diff
@@ -131,8 +154,14 @@
           _returnValue = 'NA';
         }
       } else if (v === opts_.xDomain[0]) {
+        if (data_.smallDataFlag) {
+          return '<=' + e(opts_.xDomain[1]);
+        }
         return '<=' + opts_.xDomain[1];
       } else if (v === opts_.xDomain[opts_.xDomain.length - 2]) {
+        if (data_.smallDataFlag) {
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChart.js]: This part should be merged with line 165

File: app/scripts/views/components/barChart/barChart.js [hunk 6]
```diff
@@ -363,6 +413,18 @@
         } else if (max < 100 &&
           max > 10) {
           config.divider = 2;
+        } else if (data.smallDataFlag) {
+          if (exponentRange > 20) {
+            config.divider = 8;
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChart.js]: Is the divider 8 sufficient enough if exponentRange is 100?

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: app/_components/cbio-util.js (hunk 2 above)
Comment (victoria34) [reply app/_components/cbio-util.js:694]: The default sort() should return values from small numbers to large values. But in here, it works conversely, so I have to write 'a-b'.

File: app/scripts/views/components/barChart/barChart.js (hunk 4 above)
Comment (victoria34) [reply app/scripts/views/components/barChart/barChart.js]: I'm afraid not. Line 115-120 uses linear() scale, but scientific numbers use log() scale.

File: app/_components/cbio-util.js (hunk 1 above)
Comment (victoria34) [reply app/_components/cbio-util.js]: Good Idea. Thanks!

File: app/scripts/views/components/barChart/barChart.js (hunk 5 above)
Comment (victoria34) [reply app/scripts/views/components/barChart/barChart.js]: var e = d3.format('.1e');// convert small data to scientific notation format
For scientific numbers, they should be covert to scientific format by e(). If I merged them, normal numbers will also show in scientific format.

File: app/_components/cbio-util.js (hunk 3 above)
Comment (victoria34) [reply app/_components/cbio-util.js]: You're right. I will use a for loop to make sure min and max values are valid numbers.

File: app/_components/cbio-util.js (hunk 2 above)
Comment (zhx828) [reply app/_components/cbio-util.js:694]: a-b is from small to large values

File: app/scripts/views/components/barChart/barChart.js (hunk 4 above)
Comment (zhx828) [reply app/scripts/views/components/barChart/barChart.js]: sorry 106-107

File: app/scripts/views/components/barChart/barChart.js (hunk 5 above)
Comment (zhx828) [reply app/scripts/views/components/barChart/barChart.js]: You still need the if condition, but parts `return '>' + ` and `opts_.xDomain[opts_.xDomain.length - 3]` are the same

File: app/scripts/views/components/barChart/barChart.js (hunk 6 above)
Comment (victoria34) [reply app/scripts/views/components/barChart/barChart.js]: if (data.smallDataFlag) {
          var numberOfTickValues = 4;
          config.divider = Math.round(exponentRange / numberOfTickValues);
}
Now config.divider is calculated automatically.

File: app/_components/cbio-util.js (hunk 2 above)
Comment (victoria34) [reply app/_components/cbio-util.js:694]: I didn't explain clearly. Values.sort() in here return data based on kind of alphabetical order, so I have to mention I want to sort data in ascending order.

File: app/scripts/views/components/barChart/barChart.js (hunk 4 above)
Comment (victoria34) [reply app/scripts/views/components/barChart/barChart.js]: if (logScale || data_.smallDataFlag) {
        chartInst_.x(d3.scale.log().nice()
          .domain([opts_.minDomain, opts_.maxDomain]));
 }
Sure, I have merged them.

File: app/scripts/views/components/barChart/barChart.js (hunk 5 above)
Comment (victoria34) [reply app/scripts/views/components/barChart/barChart.js]: formattedValue = opts_.xDomain[1];
        if (data_.smallDataFlag) {
          return '<=' + e(formattedValue);
        }
        return '<=' + formattedValue;

**Pull request closed without merging**